src/impl_windows.rs

```diff
@@ -43,11 +40,55 @@ pub fn runas_impl(cmd: &Command) -> io::Result<ExitStatus> {
         .collect::<Vec<_>>();
 
     unsafe {
-        let show = if cmd.hide { 0 } else { 1 };
-        Ok(mem::transmute(rust_win_runas(
-            file.as_ptr(),
-            params.as_ptr(),
-            show,
-        )))
+        use winapi::um::combaseapi::{CoInitializeEx, CoUninitialize};
+        use winapi::um::objbase::{COINIT_APARTMENTTHREADED, COINIT_DISABLE_OLE1DDE};
+        use winapi::um::shellapi::{SHELLEXECUTEINFOW, ShellExecuteExW, SEE_MASK_NOASYNC, SEE_MASK_NOCLOSEPROCESS};
+        use winapi::um::winbase::INFINITE;
+        use winapi::um::winuser::{SW_HIDE, SW_NORMAL};
+        use winapi::shared::winerror::SUCCEEDED;
+        use winapi::shared::minwindef::{DWORD, TRUE};
+        use winapi::um::synchapi::WaitForSingleObject;
+        use winapi::um::processthreadsapi::GetExitCodeProcess;
+        use std::ptr;
+
+        let show = if cmd.hide { SW_HIDE } else { SW_NORMAL };
+
+        let mut code : DWORD = 0;
+        let mut sei = SHELLEXECUTEINFOW { 
+            cbSize: mem::size_of::<SHELLEXECUTEINFOW>() as u32,
+            fMask: SEE_MASK_NOASYNC | SEE_MASK_NOCLOSEPROCESS,
+            lpVerb: (*"runas\0").encode_utf16().collect::<Vec<u16>>().as_ptr(),
+            lpFile: file.as_ptr(),
+            lpParameters: params.as_ptr(),
+            nShow: show,
+            dwHotKey: 0,
+            hInstApp: ptr::null_mut(),
+            hMonitor: ptr::null_mut(),
+            hProcess: ptr::null_mut(),
+            hkeyClass: ptr::null_mut(),
+            hwnd: ptr::null_mut(),
+            lpClass: ptr::null_mut(),
+            lpDirectory: ptr::null_mut(),
+            lpIDList: ptr::null_mut(),
+        };
+
+        let coinitializeex_result = CoInitializeEx(ptr::null_mut(), COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
+        
+        let mut succesfull = ShellExecuteExW(&mut sei) == TRUE && sei.hProcess != ptr::null_mut();
+
+        if succesfull {
+            WaitForSingleObject(sei.hProcess, INFINITE);
+            succesfull = GetExitCodeProcess(sei.hProcess, &mut code) != 0;
+        }
+        
+        if SUCCEEDED(coinitializeex_result) {
+            CoUninitialize();
+        }
+        
+        if succesfull {
+            Ok(ExitStatus::from_raw(code))
+        } else {
+            Err(std::io::Error::last_os_error())
```

Comment:
One more thing: this call to `last_os_error()` will most likely get the error status from `CoUninitialize()` (which, I suspect, would most likely cause this pattern to match the error returned: `Err(err) if err.success()`).

I've also found in my own testing of the function, getting the last error immediately after calling `ShellExecuteExW` is also not very useful, as it always returns an error that indicates the user cancelled the action. Instead, the function sets the value of `sei.hInstApp` to an error code that is, unfortunately, custom for this particular function.

~`std::io::Error::from_raw_os_code(-1)` would, I think, be the equivalent of the current behavior on master.~

After looking more closely, the current master behavior is to always return `Ok(status)` no matter what, where the *actual* indication of failure is when `!status.success()`. That seems somewhat odd to me, as to me it would imply the process always runs, which isn't guaranteed (ie, passing an invalid path would effectively return `Ok(-1)` rather than `Err(file_not_found)`).

So I think the equivalent of current master would be `Ok(ExitStatus::from_raw(-1))`

I'm not sure what the behavior is exactly on other platforms, so I'm not sure what the right answer is here to make it consistent. But if it helps at all, this is a snippet of how I've defined errors in a shell execute wrapper I'd been working on: https://gist.github.com/TyPR124/03f9d3d489d1726bb2cf8fad760b5633 (note: the `NullInInput` variant is not an actual Windows error, but specific to my wrapper)

Reply:
It is indeed possible that `CoUninitialize` could/would override the `last_os_error` however that would require some testing, I don't use the code produced by this `PR` as I've decided to go a different route for my own applications where I believe mapping it to [std::process::Child](https://doc.rust-lang.org/std/process/struct.Child.html) is more worthwhile, giving you all the possibilities the `Child` struct gives you. See my [main](https://github.com/SonnyX/rust-runas/tree/master/) branch.

As for the `Ok(ExitStatus::from_raw(-1))`, this seems to be a mistake by the owner of the crate, as `-1` is not even a valid exit code in windows, as they are supposed to be `u32`. If anything, providing better error codes would probably be beneficial to the users of this crate.